### PR TITLE
initial support for bare repositories

### PR DESCRIPTION
**Pull request by savely-krasovsky**
This PR adds support for bare repositories. It's especially useful to use with pre-receive hooks. Currently it's a draft, at least because go-git has a bug in upstream which doesn't allow to scan incoming commits in pre-receive hooks.

### CODE REVIEW

**Comment by L11R** [comment]
go-git changes also under review currently: https://github.com/go-git/go-git/pull/812

File: go.mod [hunk 1]
```diff
@@ -2,7 +2,10 @@ module github.com/trufflesecurity/trufflehog/v3
 
 go 1.20
 
-replace github.com/jpillora/overseer => github.com/trufflesecurity/overseer v1.1.7-custom5
+replace (
+	github.com/go-git/go-git/v5 v5.7.0 => ../go-git
```
Comment (zricethezav) [inline go.mod:6]: @L11R **suggestion (blocking)**: This will need to be removed once the bug in go-git has been resolved

File: pkg/gitparse/gitparse.go [hunk 2]
```diff
@@ -134,7 +135,24 @@ func (c *Parser) RepoPath(ctx context.Context, source string, head string, abbre
...
 	absPath, err := filepath.Abs(source)
 	if err == nil {
-		cmd.Env = append(cmd.Env, fmt.Sprintf("GIT_DIR=%s", filepath.Join(absPath, ".git")))
+		if !isBare {
+			cmd.Env = append(cmd.Env, fmt.Sprintf("GIT_DIR=%s", filepath.Join(absPath, ".git")))
+		} else {
+			cmd.Env = append(cmd.Env, fmt.Sprintf("GIT_DIR=%s", absPath))
+
+			// We need those variables to handle incoming commits
+			// while using trufflehog in pre-receive hooks
```
Comment (zricethezav) [inline pkg/gitparse/gitparse.go:144]: @L11R **suggestion**: Rather than iterating over all the environment variables we should attempt to add them to `cmd.Env` with `os.Getenv`

**Comment by zricethezav** [comment]
Adding Draft label until the issue with go-git has been resolved

File: go.mod (hunk 1 above)
Comment (L11R) [reply go.mod:6]: @zricethezav sure!

File: pkg/gitparse/gitparse.go (hunk 2 above)
Comment (L11R) [reply pkg/gitparse/gitparse.go]: @zricethezav should be better now! Also replaced fmt.Sprintf with simple concatenation, it should be more performant.

File: go.mod (hunk 1 above)
Comment (L11R) [reply go.mod:6]: My fix in go-git has merged, waiting a release.

**Comment by L11R** [comment]
@zricethezav go-git team has merged my fix. In my opinion gitparse_test.go has not full .git repository tests. We need fixtures with usual and bare repos to test against them.

File: README.md [hunk 3]
```diff
@@ -283,6 +283,53 @@ repos:
       stages: ["commit", "push"]
 ```
 
+# Pre-receive hook
```
Comment (zricethezav) [inline README.md:286]: @L11R _I think_ this is something we don't want in our README, _yet_.  We haven't tested it yet and don't want to showcase something that hasn't been thoroughly tested.

Also this README addition is unrelated to this PR. The "Pre-commit" typos are fine though

Comment (L11R) [reply README.md:286]: @zricethezav sure, understand. Will remove!

**Comment by L11R** [comment]
Oh, go-git team released v5.8.0 with my fix, merged changes to `go.mod` and `go.sum` accordingly.

**Comment by dustin-decker** [comment]
I just merged go-git v5.8.1 into TruffleHog @L11R

**Comment by L11R** [comment]
@dustin-decker nice! Can someone from your team review it again? I fixed issues with doc and dependencies.

**Review by mcastorina — COMMENTED** [review]
Thanks for the improvement! I spot-tested the changes locally and it didn't scan the bare repository. Did I do something wrong?

```
» git clone --bare https://github.com/trufflesecurity/test_keys /tmp/test_keys.git
Cloning into bare repository '/tmp/test_keys.git'...
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 13 (delta 0), reused 0 (delta 0), pack-reused 9
Receiving objects: 100% (13/13), done.
```
```
» ./trufflehog --no-update git --bare file:///tmp/test_keys.git
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

2023-08-03T08:55:14-05:00       info-0  trufflehog      finished scanning       {"chunks": 0, "bytes": 0, "verified_secrets": 0, "unverified_secrets": 0}
```

Additionally, I had to update `pkg/engine/git.go` as the automatic merge caused a Go compilation error.

**Comment by L11R** [comment]
@mcastorina I guess the problem relates to recent git scan refactoring. Will research

**Comment by L11R** [comment]
@mcastorina I've prepared a new patch, now it should work again. It was definitively after merging this: #1518

**Review by mcastorina — APPROVED** [review]
lgtm, nice work!
